File: 08. Trees/02. Binary Search Tree/04. Delete node in a BST/main.js
class TreeNode {
    constructor(val) {
        this.val = val;
        this.left = this.right = null;
    }
}

// Find minimum node in a subtree
function findMin(root) {
    while (root.left) root = root.left;
    return root;
}

// Recursive Delete
function deleteRecursive(root, key) {
    if (!root) return root;

    if (key < root.val) root.left = deleteRecursive(root.left, key);
    else if (key > root.val) root.right = deleteRecursive(root.right, key);
    else {
        if (!root.left) return root.right;
        if (!root.right) return root.left;

        let temp = findMin(root.right);
        root.val = temp.val;
        root.right = deleteRecursive(root.right, temp.val);
    }
    return root;
}

// Inorder Traversal
function inorder(root) {
    if (!root) return;
    inorder(root.left);
    process.stdout.write(root.val + " ");
    inorder(root.right);
}

// Main function
function main() {
    let root = new TreeNode(8);
    root.left = new TreeNode(3);
    root.right = new TreeNode(10);
    root.left.right = new TreeNode(6);
    root.right.right = new TreeNode(14);

    root = deleteRecursive(root, 10);
    process.stdout.write("Inorder after Recursive Delete: ");
    inorder(root);
    console.log();
}

main();

File: 08. Trees/02. Binary Search Tree/04. Delete node in a BST/main.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Find minimum node in a subtree
def find_min(root):
    while root.left:
        root = root.left
    return root

# Recursive Delete
def delete_recursive(root, key):
    if not root:
        return root
    
    if key < root.val:
        root.left = delete_recursive(root.left, key)
    elif key > root.val:
        root.right = delete_recursive(root.right, key)
    else:
        if not root.left:
            return root.right
        elif not root.right:
            return root.left
        
        temp = find_min(root.right)
        root.val = temp.val
        root.right = delete_recursive(root.right, temp.val)
    
    return root

# Inorder Traversal
def inorder(root):
    if root:
        inorder(root.left)
        print(root.val, end=" ")
        inorder(root.right)

# Main function
def main():
    root = TreeNode(8)
    root.left = TreeNode(3)
    root.right = TreeNode(10)
    root.left.right = TreeNode(6)
    root.right.right = TreeNode(14)

    root = delete_recursive(root, 10)
    print("Inorder after Recursive Delete:", end=" ")
    inorder(root)
    print()

if __name__ == "__main__":
    main()
